### DISCUSSION TITLE
Network-On-Chip Capability

### DISCUSSION THREAD

**Discussion by Soilorian** [sarchlab/mgpusim#74] [Q&A] [answered]
Hello Mr.Sun,

I have been working with the simulator a lot and wanted to know if there is anything related to Network-On-Chip already implemented in mgpu-sim? More specifically, I am looking for how the CUs, caches, and DRAM are connected to each other and also how the GPUs communicate with one another.

another question: I tried the debugging ISA option but the files it generated with `.debug` extension were all empty. I was wondering, am I missing something? I used the following command:

```sh
./fir -timing -debug-isa -gpus=1,2 -length=40960
```

and also can you guide me to where the instructions are decoded and where can I find the instructions themselves after the benchmark is compiled?

Thank you in advance!

**Comment by syifan** [comment, marked as answer]
Akita and MGPUSim provides some NOC support. See here https://github.com/sarchlab/akita/tree/v3/noc. 

Within a GPU, everything is pretty much just connected with wires (direct connections), while you can configure them to use more complex network. 

For the question regarding the debug files are empty, it is very likely to because of this line 

https://github.com/sarchlab/mgpusim/blob/v3/emu/isadebugger.go#L40
https://github.com/sarchlab/mgpusim/blob/v3/timing/cu/isa_debugger.go#L72

Remove the condition but keep the content will get all the instructions printed.

**Reply by Soilorian** [reply]
thanks for the answer

**Reply by Soilorian** [reply]
hello again mr. Sun

after a week of trying to figure out how I can use the noc you suggested, I failed to match together all of the different classes and methods in your pre-made NoC. Is there a sample project or maybe another repo to help me understand how can I use this noc system? or maybe a more complete explanation than the comments provided in the code?

**Reply by syifan** [reply]
Could you please let me know what you want to build? Perhaps I can help you configure such systems.

**Reply by Soilorian** [reply]
I am aiming to implement the [SAC](https://dl.acm.org/doi/abs/10.1145/3579371.3589078)(share-aware cache) cache system for LLC of the gpus. in order to do so, I need to be able to decide that the LLC caches data going to the SM from local or remote sources or it will cache the data coming from the gpu's memory for remote or local accesses.

![image](https://github.com/sarchlab/mgpusim/assets/109273848/2af9c99f-0dc2-4d14-8ad6-85ec88d57216)

this image shows exactly what I'm trying to achieve. 

![image](https://github.com/sarchlab/mgpusim/assets/109273848/68613d7b-c73d-4d0c-9e2d-d39af6dbfed1)

and this is what my final project should look like.

also I think you have missed my question regarding how can I find the instructions being executed for the program. I need them for the pre-process needed for SAC.

thanks in advance!

**Reply by syifan** [reply]
OK. I would say this is not really an "NOC" problem but a memory system configuration problem since the NOC latency is not the most critical thing you care about, but the topology, right? The default configuration is pretty much A. Have you checked how the current L1 cache is connected with the L2 cache? Here is the code. 
https://github.com/sarchlab/mgpusim/blob/v3/samples/runner/r9nanobuilder.go#L297. 

If you understand this code, it may help you configure (b) and (c). For the NOC in the figure, we pretty much just use a direct connection, which serves as a crossbar.

**Reply by Soilorian** [reply]
yes that's true, the latency of NOC is not of high importance to me at the moment.

I ,in fact, did take a look at this code but I didn't realize how each GPU can access other GPU's memories. But now that you have pointed out this line, I think that's handled by [lowModuleFinder](https://gitlab.com/akita/mem/-/blob/v3/mem/lowmodulefinder.go) and [rdmaEngine](https://github.com/sarchlab/mgpusim/blob/v3/timing/rdma/comp.go), right? so by changing their ordering and a little bit of how they function I can implement what I'm aiming for.

**Reply by syifan** [reply]
Yes. The `LowModuleFinder` is where the magic happens. RDMA is like a network interface that tunnels GPUs. To model your architecture, you definitely need to work on the `LowModuleFinder`a and `RDMA`s. It may take some effort but should be doable.
